microsetta_public_api/api/tests/test_api.py

```diff
@@ -4,6 +4,237 @@
+        _, self.client = self.build_app_test_client()
+        exp = []
+        response = self.client.get(
+            "/api/metadata/category/values/age_cat")
+        self.assertStatusCode(200, response)
+        obs = json.loads(response.data)
+        self.assertListEqual(exp, obs)
+        self.mock_method.assert_called_with(category='age_cat')
+
+    def test_metadata_category_values_returns_404(self):
+        with self.app_context():
+            self.mock_method.return_value = jsonify(
+                error=404, text='Unknown Category',
+            ), 404
+        _, self.client = self.build_app_test_client()
+        response = self.client.get(
+            "/api/metadata/category/values/age_cat")
+        self.assertStatusCode(404, response)
+        self.mock_method.assert_called_with(category='age_cat')
+
+
+class MetadataSampleIdsTests(FlaskTests):
+
+    def setUp(self):
+        super().setUp()
+        self.patcher = patch(
+            'microsetta_public_api.api.metadata.filter_sample_ids')
+        self.mock_method = self.patcher.start()
+
+    def tearDown(self):
+        self.patcher.stop()
+
+    def test_metadata_sample_ids_returns_simple(self):
+        with self.app_context():
+            self.mock_method.return_value = jsonify({
+                'sample_ids': [
+                    'sample-1',
+                    'sample-2',
+                ]
+            })
+
+        _, self.client = self.build_app_test_client()
+        exp_ids = ['sample-1', 'sample-2']
+        response = self.client.get(
+            "/api/metadata/sample-ids?age_cat=30s&bmi_cat=normal")
+        self.assertStatusCode(200, response)
+        obs = json.loads(response.data)
+        self.assertCountEqual(['sample_ids'], obs.keys())
+        self.assertCountEqual(obs['sample_ids'], exp_ids)
+        self.mock_method.assert_called_with(age_cat='30s', bmi_cat='normal')
+
+    def test_metadata_sample_ids_returns_empty(self):
+        with self.app_context():
+            self.mock_method.return_value = jsonify({
+                'sample_ids': [
+                ]
+            })
+
+        _, self.client = self.build_app_test_client()
+        response = self.client.get(
+            "/api/metadata/sample-ids?age_cat=30s&bmi_cat=normal")
+        self.assertStatusCode(200, response)
+        obs = json.loads(response.data)
+        self.assertCountEqual(['sample_ids'], obs.keys())
+        self.assertEqual(obs['sample_ids'], [])
+
+    def test_metadata_sample_ids_returns_extra_categories(self):
+        with self.app_context():
+            self.mock_method.return_value = jsonify({
+                'sample_ids': [
+                    'sample-1',
+                ],
+                'some-extra-cat': ['stuff'],
+            })
+
+        _, self.client = self.build_app_test_client()
+        response = self.client.get(
+            "/api/metadata/sample-ids?age_cat=30s&bmi_cat=normal")
+        self.assertStatusCode(500, response)
```

Comment:
We really shouldn't ever be throwing `500`s... is this something expected that can happen?

Reply:
My understanding is that if your operation gives a response that does not follow the schema for your API, the client will give a 500 error. So these types of checks for 500's are generally verifying that certain types of outputs are disallowed by my API spec.